src/beast/beast/container/vectum.h

```diff
@@ -0,0 +1,642 @@
+    {
+        return ptr_ != rhs.ptr_;
+    }
+
+    template <bool maybe_const>
+    bool
+    operator< (basic_iterator<
+        maybe_const> const& rhs) const
+    {
+        return ptr_ < rhs.ptr_;
+    }
+
+    template <bool maybe_const>
+    bool
+    operator> (basic_iterator<
+        maybe_const> const& rhs) const
+    {
+        return ptr_ > rhs.ptr_;
+    }
+
+    template <bool maybe_const>
+    bool
+    operator<= (basic_iterator<
+        maybe_const> const& rhs) const
+    {
+        return ptr_ <= rhs.ptr_;
+    }
+
+    template <bool maybe_const>
+    bool
+    operator>= (basic_iterator<
+        maybe_const> const& rhs) const
+    {
+        return ptr_ >= rhs.ptr_;
+    }
+
+    basic_iterator&
+    operator++()
+    {
+        ptr_ = increment(1);
+        return *this;
+    }
+
+    basic_iterator
+    operator++ (int) const
+    {
+        basic_iterator _(*this);
+        ++(*this);
+        return _;
+    }
+
+    basic_iterator&
+    operator--()
+    {
+        ptr_ = increment(-1);
+        return *this;
+    }
+
+    basic_iterator
+    operator-- (int)
+    {
+        basic_iterator _(*this);
+        --(*this);
+        return _;
+    }
+
+    basic_iterator&
+    operator+= (std::ptrdiff_t n)
+    {
+        ptr_ = increment(n);
+        return *this;
+    }
+
+    basic_iterator&
+    operator-= (std::ptrdiff_t n)
+    {
+        ptr_ = increment(-n);
+        return *this;
+    }
+
+    friend
+    basic_iterator
+    operator+ (std::ptrdiff_t n,
+        basic_iterator const& rhs)
+    {
+        basic_iterator _(rhs);
+        _ += n;
+        return _;
+    }
+
+    friend
+    basic_iterator
+    operator+ (basic_iterator const& lhs,
+        std::ptrdiff_t n)
+    {
+        return basic_iterator(
+            lhs.increment(n), lhs.aligned_size_);
+    }
+
+    friend
+    basic_iterator
+    operator- (basic_iterator const& lhs,
+        std::ptrdiff_t n)
+    {
+        return basic_iterator(
+            lhs.increment(-n), lhs.aligned_size_);
+    }
+
+    friend
+    std::ptrdiff_t
+    operator- (basic_iterator const& lhs,
+                basic_iterator const& rhs)
+    {
+        assert(lhs.aligned_size_ == rhs.aligned_size_);
+        return
+            (reinterpret_cast<byte_t>(lhs.ptr_) -
+                reinterpret_cast<byte_t>(rhs.ptr_)) /
+                    lhs.aligned_size_;
+    }
+
+private:
+    friend class vectum;
+
+    basic_iterator (ptr_t ptr,
+            std::size_t aligned_size)
+        : ptr_ (ptr)
+        , aligned_size_ (aligned_size)
+    {
+    }
+
+    ptr_t
+    increment (std::ptrdiff_t n) const
+    {
+        return reinterpret_cast<ptr_t> (
+            (reinterpret_cast<byte_t>(ptr_) +
+                n * aligned_size_));
+    }
+};
+
+//------------------------------------------------------------------------------
+
+template <class T>
+vectum<T>::vectum (size_type value_size)
+    : value_size_ (value_size)
+    , aligned_size_ (((value_size +
```

Comment:
Honestly, I'd love to see unit tests for various gnarly types - `char` and `short` and `long long` and some struct that's 12 bytes in size...